kernel/src/arch/x86_64/pci.rs

```diff
@@ -0,0 +1,214 @@
+use super::port;
+use crate::driver;
+use alloc::vec::Vec;
+
+const CONFIG_ADDRESS: u16 = 0xCF8;
+const CONFIG_DATA: u16 = 0xCFC;
+const ENABLE: u32 = 1 << 31;
+
+const VENDOR_ID: u8 = 0;
+pub const COMMAND: u8 = 4;
+const CLASS: u8 = 10;
+const BAR: u8 = 16;
+
+#[derive(Debug, Clone, Copy)]
+pub struct Device {
+    pub bus: u8,
+    pub slot: u8,
+}
+
+impl Device {
+    /// Creates a PCI device at bus and slot address
+    /// Returns None if device does not exist
+    pub fn new(bus: u8, slot: u8) -> Option<Device> {
+        let device = Device { bus, slot };
+        match device.exists(0) {
```

Comment:
A PCI device can have up to 8 functions, what they do is device dependent however,
for instance these are the functions of the Gigabit Ethernet Controllers family
![image](https://user-images.githubusercontent.com/31779371/133796983-74eb9898-d885-4a50-b3c6-730de52cc240.png)
Meaning i can't do much more than create an enum Function0, Function1, etc which doesn't carry more information than just the value itself.
Furthermore, the reason why 0 is used almost everywhere is because it is guaranteed to be supported by an existing PCI device.